attest/wrapped_tpm20.go

```diff
@@ -171,27 +180,95 @@ func (t *wrappedTPM20) getStorageRootKeyHandle(parent ParentKeyConfig) (tpmutil.
 	default:
 		return 0, false, fmt.Errorf("unsupported SRK algorithm: %v", parent.Algorithm)
 	}
-	keyHnd, _, err := tpm2.CreatePrimary(t.rwc, tpm2.HandleOwner, tpm2.PCRSelection{}, "", "", srkTemplate)
+	err = t.createEK(srkTemplate, srkHandle, rerr)
 	if err != nil {
-		return 0, false, fmt.Errorf("ReadPublic failed (%v), and then CreatePrimary failed: %v", rerr, err)
+		return 0, false, err
 	}
-	defer tpm2.FlushContext(t.rwc, keyHnd)
+	return srkHandle, true, nil
+}
 
-	err = tpm2.EvictControl(t.rwc, "", tpm2.HandleOwner, keyHnd, srkHandle)
-	if err != nil {
-		return 0, false, fmt.Errorf("EvictControl failed: %v", err)
+func (t *wrappedTPM20) getKeyHandleKeyMap() (map[crypto.PublicKey]tpmutil.Handle, error) {
+	rvalue := make(map[crypto.PublicKey]tpmutil.Handle)
+	// NOTE: this list should be replaced by a call to an
+	// equivalent of:  "tpm2_getcap handles-persistent"
+	keyHandlesToSearch := []tpmutil.Handle{
+		commonRSAEkEquivalentHandle,
+		commonECCEkEquivalentHandle,
+		commonECCEkEquivalentHandle + 1,
+		commonECCEkEquivalentHandle + 2,
 	}
-
-	return srkHandle, true, nil
+	for _, keyHandle := range keyHandlesToSearch {
+		pub, _, _, err := tpm2.ReadPublic(t.rwc, keyHandle)
+		if err != nil {
+			continue
+		}
+		if pub.RSAParameters != nil || pub.ECCParameters != nil {
+			key, err := pub.Key()
+			if err != nil {
+				return nil, err
+			}
+			rvalue[key] = keyHandle
+		}
+	}
+	return rvalue, nil
 }
 
 func (t *wrappedTPM20) ekCertificates() ([]EK, error) {
 	var res []EK
-	if rsaCert, err := readEKCertFromNVRAM20(t.rwc, nvramRSACertIndex); err == nil {
-		res = append(res, EK{Public: crypto.PublicKey(rsaCert.PublicKey), Certificate: rsaCert, handle: commonRSAEkEquivalentHandle})
+	certIndexes := []int{nvramRSACertIndex, nvramECCCertIndex, nvram3KRSACertIndex, nvramP384CertIndex}
+	keyHandleMap, err := t.getKeyHandleKeyMap()
+	if err != nil {
+		return nil, err
 	}
-	if eccCert, err := readEKCertFromNVRAM20(t.rwc, nvramECCCertIndex); err == nil {
-		res = append(res, EK{Public: crypto.PublicKey(eccCert.PublicKey), Certificate: eccCert, handle: commonECCEkEquivalentHandle})
+	for _, certIndex := range certIndexes {
+		if cert, err := readEKCertFromNVRAM20(t.rwc, tpmutil.Handle(certIndex)); err == nil {
+			var handleToUse tpmutil.Handle
+			keyfound := false
+		FindKeyHandleLoop:
+			//This section finds the keyhandle
+			for key, keyHandle := range keyHandleMap {
```

Comment:
So this code is iterating over the map, which somewhat defeats the purpose of a map. Because (I believe) crypto.PublicKEy will be compared using pointer comparison, we can't use this type directly. So instead it could be converted to a string first.

func keyToMapKey(pubKey crypto.PublicKey) (string, error) {
    // 1. Marshal the public key into its DER-encoded byte form (PKIX standard)
    derBytes, err := x509.MarshalPKIXPublicKey(pubKey)
    if err != nil {
        return "", err
    }
    // 2. Convert the unique byte slice to a string for map key use
    return string(derBytes), nil
}

This could then be used to avoid the looping and even the type switch possibly. This will clear up the code more.